datastore/additional_data/sources/nspl.py

```diff
@@ -0,0 +1,166 @@
+
+        for file in zip_file.filelist:
+            if not file.filename.endswith(".csv") or not file.filename.startswith(
+                "Data/multi_csv/NSPL_"
+            ):
+                continue
+
+            print("[postcodes] Opening %s" % file.filename)
+
+            postcode_count = 0
+
+            with zip_file.open(file, "r") as postcode_csv:
+                postcode_csv = io.TextIOWrapper(postcode_csv)
+                reader = csv.DictReader(postcode_csv)
+
+                for record in reader:
+                    # null any blank fields (or ones with a dummy code in)
+                    for key in record:
+                        if record[key] == "" or record[key] in [
+                            "E99999999",
+                            "S99999999",
+                            "W99999999",
+                            "N99999999",
+                        ]:
+                            record[key] = None
+
+                    # date fields
+                    for field in ["dointr", "doterm"]:
+                        if record[field]:
+                            date_field = datetime.strptime(record[field], "%Y%m")
+                            record[field] = str(date_field)
+
+                    # latitude and longitude
+                    for field in ["lat", "long"]:
+                        if record[field]:
+                            record[field] = float(record[field])
+                            if record[field] == 99.999999:
+                                record[field] = None
+                    if record["lat"] and record["long"]:
+                        record["location"] = {
+                            "lat": record["lat"],
+                            "lon": record["long"],
+                        }
+
+                    # integer fields
+                    for field in [
+                        "oseast1m",
+                        "osnrth1m",
+                        "usertype",
+                        "osgrdind",
+                        "imd",
+                    ]:
+                        if record[field]:
+                            record[field] = int(record[field])
+
+                    # add postcode hash
+                    record["hash"] = hashlib.md5(
+                        record["pcds"].lower().replace(" ", "").encode()
+                    ).hexdigest()
+
+                    postcodes.append(record)
+                    postcode_count += 1
+                print("[postcodes] Processed %s postcodes" % postcode_count)
+
+        return postcodes
+
+    def save_nspl_data(self, data):
+        for record in data:
+            nspl_postcode = (
+                record["pcd"]
+                if record["pcd"]
+                else record["pcd2"]
+                if record["pcd2"]
+                else record["pcds"]
+            )
+            postcode = "".join(nspl_postcode.split()).upper()
+            NSPL.objects.create(postcode=postcode, data=record).save()
```

Comment:
What happens the second time the data is loaded? Does it overwrite the old ones or add to the table?

Reply:
It overwrites them. See line 130:
https://github.com/ThreeSixtyGiving/datastore/pull/34/files#diff-91824db8d74ba40fbf27b3c9b6380dd3R130

Reply:
Anywhere there is an `objects.create` in a for loop we could really just use bulk_create , I think this could take a very long time otherwise. https://docs.djangoproject.com/en/3.0/ref/models/querysets/#bulk-create